subsys/secure_storage/src/its/transform/aead_get.c

```diff
@@ -0,0 +1,114 @@
+/* Copyright (c) 2024
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/secure_storage/its/transform/aead_get.h>
+#include <zephyr/drivers/hwinfo.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+#include <psa/crypto.h>
+#include <string.h>
+#include <sys/types.h>
+
+LOG_MODULE_DECLARE(secure_storage, CONFIG_SECURE_STORAGE_LOG_LEVEL);
+
+#ifdef CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER_DEFAULT
+
+void secure_storage_its_transform_aead_get_scheme(psa_key_type_t *key_type, psa_algorithm_t *alg)
+{
+	*key_type = PSA_KEY_TYPE_AES;
+	*alg = PSA_ALG_GCM;
+}
+#endif /* CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER_DEFAULT */
+
+#ifdef CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER_DEFAULT
+
+#ifndef CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NO_INSECURE_KEY_WARNING
+
+static int warn_insecure_key(void)
+{
+	const char *warning = "The default (insecure) PSA ITS encryption key provider is in use.";
+
+	printk("WARNING: %s\n",  warning);
+	LOG_WRN("%s", warning);
+	return 0;
+}
+SYS_INIT(warn_insecure_key, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+#endif /* !CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NO_INSECURE_KEY_WARNING */
+
+psa_status_t secure_storage_its_transform_aead_get_key(secure_storage_its_uid_t uid,
+		uint8_t key[static CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE])
+{
+	struct {
+		secure_storage_its_uid_t uid;
+		uint8_t device_id[8];
+	} __packed hash_input = {
+		.uid = uid,
+	};
+	size_t hash_input_len = sizeof(hash_input.uid);
+	size_t hash_len;
+
+#ifdef CONFIG_HWINFO
+	ssize_t hwinfo_ret;
+
+	hwinfo_ret = hwinfo_get_device_eui64(hash_input.device_id);
+	if (hwinfo_ret == 0) {
+		hash_input_len += 8;
+	} else {
+		hwinfo_ret = hwinfo_get_device_id(hash_input.device_id,
+						  sizeof(hash_input.device_id));
+		if (hwinfo_ret > 0) {
+			hash_input_len += hwinfo_ret;
+		}
+	}
+#endif /* CONFIG_HWINFO */
+
+	enum { HASH_OUTPUT_LEN = PSA_HASH_LENGTH(PSA_ALG_SHA_256) };
+	BUILD_ASSERT(CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE <= HASH_OUTPUT_LEN);
+
+	if (CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_SIZE == HASH_OUTPUT_LEN) {
+		return psa_hash_compute(PSA_ALG_SHA_256, (uint8_t *)&hash_input, hash_input_len,
```

Comment:
Ok, thanks for the clarification. Perhaps my point-of-view is not too "embedded", but a SHA-256 is only 32 bytes so I didn't thought that buffer's length could be that bad.

Reply:
It wouldn't be too bad by itself, but the stack is already under heavy usage because of the other, bigger buffers, so any possible saving is good to take.